elements/pfe-avatar/src/pfe-avatar.js

```diff
@@ -0,0 +1,352 @@
+        }
+      }
+    }
+  }
+
+  /**
+   * Draw a square at the given position, mirrored onto both the left and right half of the canvas.
+   */
+  _drawMirroredSquare(x, y) {
+    if (this._ctx) {
+      this._drawSquare(x, y);
+      this._drawSquare(7 - x, y);
+    }
+  }
+
+  _drawSquare(x, y) {
+    this._ctx.fillRect(
+      this._squareSize * x,
+      this._squareSize * y,
+      this._squareSize,
+      this._squareSize
+    );
+  }
+
+  _drawTrianglePattern(pattern) {
+    this._ctx.fillStyle = this.color2;
+    if (this._ctx) {
+      let i = pattern.length;
+      while (i--) {
+        if (pattern[i]) {
+          const x = Math.floor(i / 2) % 2;
+          const y = Math.floor(i / 4);
+          const alt = i % 4;
+
+          const p1 = [x, y];
+          const p2 = [x, y];
+          const p3 = [x, y];
+
+          switch (alt) {
+            case 0:
+              p2[1]++;
+              p3[0]++;
+              p3[1]++;
+              break;
+            case 1:
+              p2[0]++;
+              p3[0]++;
+              p3[1]++;
+              break;
+            case 2:
+              p2[0]++;
+              p3[1]++;
+              break;
+            case 3:
+              p1[0]++;
+              p2[0]++;
+              p2[1]++;
+              p3[1]++;
+              break;
+          }
+
+          this._drawMirroredTriangle(p1, p2, p3);
+        }
+      }
+    }
+  }
+
+  /**
+   * Draw a square at the given position in the top-left quadrant of the
+   * canvas, and mirrored to the other three quadrants.
+   */
+  _drawMirroredTriangle(p1, p2, p3) {
+    if (this._ctx) {
+      this._drawTriangle(p1, p2, p3);
+      this._drawTriangle(
+        [4 - p1[0], p1[1]],
+        [4 - p2[0], p2[1]],
+        [4 - p3[0], p3[1]]
+      );
+    }
+  }
+
+  _drawTriangle(p1, p2, p3) {
+    this._ctx.beginPath();
+    this._ctx.moveTo(...p1.map(c => c * this._triangleSize));
+    this._ctx.lineTo(...p2.map(c => c * this._triangleSize));
+    this._ctx.lineTo(...p3.map(c => c * this._triangleSize));
+    this._ctx.closePath();
+    this._ctx.fill();
+    this._ctx.fill();
+  }
+
+  _drawGradient() {
+    const gradient = this._ctx.createLinearGradient(
+      0,
+      this._canvas.height,
+      this._canvas.width,
+      0
+    );
+    const color = this.color2;
+    let gradientColor1 = color;
+    let gradientColor2 = color;
+    if (/^#[A-f0-9]{3}$/.test(color)) {
+      // color is of the form "#fff"
+      gradientColor1 += "c";
+      gradientColor2 += "0";
+    } else if (/^#[A-f0-9]{6}$/.test(color)) {
+      // color is of the form "#ffffff"
+      gradientColor1 += "cc";
+      gradientColor2 += "00";
+    }
+    gradient.addColorStop(0, gradientColor1);
+    gradient.addColorStop(1, gradientColor2);
+    gradient.addColorStop(1, gradientColor1);
+    this._ctx.fillStyle = gradient;
+    this._ctx.fillRect(0, 0, this._canvas.width, this._canvas.height);
+  }
+}
+
+PfeAvatar._registerColors();
```

Comment:
Curious why do this outside of the element declaration?

Reply:
It's tied to the class, not an individual element, so it can't go in the constructor or connectedCallback.  It needs to be called only once and applies to all pfe-avatars on the page.

Reply:
Oh very clever!  Minor question, since the function is designed to be run from outside the element's definitions, maybe it shouldn't be prefixed with the underscore?  I understood the underscore to be scoped to internal-only functions.

Reply:
It isn't intended to be run from outside this file, hence the underscore.